### PR TITLE
Fix+improve Commit Status API

### PR DESCRIPTION
**Pull request by cvrebert**
* Fixes same bug as #1102
* Use `Optional`
* Properly test JSONification of `RtStatus`
* Move JSONification of `RtStatus` into `RtStatus`
* Make `Status.State` a `StringEnum`

### CODE REVIEW

**Comment by dmarkov** [comment]
@cvrebert Thanks for the pull request, let me find a reviewer..

**Comment by dmarkov** [comment]
@caarlos0 review this one please

File: src/main/java/com/jcabi/github/RtStatus.java [hunk 1]
```diff
@@ -40,7 +43,6 @@
  * @version $Id$
  * @since 0.23
  */
-@Immutable
```
Comment (caarlos0) [inline src/main/java/com/jcabi/github/RtStatus.java]: @cvrebert why was this removed?

File: src/main/java/com/jcabi/github/Status.java [hunk 2]
```diff
@@ -52,57 +53,82 @@
...
         /**
          * Error state.
          */
-        Error,
+        Error("error"),
         /**
          * Failure state.
          */
-        Failure;
+        Failure("failure");
```
Comment (caarlos0) [inline src/main/java/com/jcabi/github/Status.java]: @cvrebert as far as I know is a convention to make all enum items in uppercase...

File: src/main/java/com/jcabi/github/Status.java [hunk 3]
```diff
@@ -52,57 +53,82 @@
...
+        State(final String stat) {
+            this.state = stat;
+        }
+
+        /**
+         * Get enum value from identifier string.
+         * @param ident Commit status state string
+         * @return Corresponding State
+         */
+        public static State forValue(
```
Comment (caarlos0) [inline src/main/java/com/jcabi/github/Status.java]: @cvrebert let's simplify this method:

```java
State.valueOf(ident.toUpperCase())
```

(assuming my previous comment)

File: src/main/java/com/jcabi/github/Status.java [hunk 4]
```diff
@@ -52,57 +53,82 @@
...
         }
+
+        /**
+         * Get commit status state string.
+         * @return String
+         */
+        @Override
+        @NotNull(message = "identifier is never NULL")
+        public String identifier() {
+            return this.state;
```
Comment (caarlos0) [inline src/main/java/com/jcabi/github/Status.java]: @cvrebert this is just the name of the enum item in lowercase... can't we just return that?

File: src/test/java/com/jcabi/github/RtStatusTest.java [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            new RtStatus(
+                Status.State.Pending,
+                Optional.of(url),
+                Optional.of(pending),
+                Optional.<String>absent()
+            ).json().toString(),
+            // @checkstyle LineLength (1 line)
+            Matchers.equalTo("{\"state\":\"pending\",\"target_url\":\"http://status.jcabi-github.invalid/42\",\"description\":\"Kragle is drying...\"}")
+        );
+    }
```
Comment (caarlos0) [inline src/test/java/com/jcabi/github/RtStatusTest.java]: @cvrebert this test is actually testing four thing. Let's break it up accordingly...

**Comment by caarlos0** [comment]
@cvrebert couple of comments above.

File: src/main/java/com/jcabi/github/RtStatus.java (hunk 1 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/RtStatus.java]: @caarlos0 The immutability checker thinks that `Optional` is mutable for some reason, and this makes compilation fail.

File: src/main/java/com/jcabi/github/Status.java (hunk 4 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/Status.java]: @caarlos0 I prefer this because it's more explicit and doesn't involve messing with Locales.

File: src/main/java/com/jcabi/github/Status.java (hunk 2 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/Status.java]: @caarlos0 Yeah, IDK why the old code was allowed in like this... I'll add a commit to change it.

File: src/main/java/com/jcabi/github/Status.java (hunk 3 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/Status.java]: @caarlos0 Again, I prefer to be explicit and avoid this sort of magic and dependency on Locale (for case conversion).

Comment (caarlos0) [inline src/main/java/com/jcabi/github/Status.java]: @cvrebert for that we can pass `Locale.ENGLISH` to it...

File: src/main/java/com/jcabi/github/RtStatus.java (hunk 1 above)
Comment (caarlos0) [inline src/main/java/com/jcabi/github/RtStatus.java]: @cvrebert weird... maybe a bug in jcabi-aspects...

cc/ @yegor256

File: src/test/java/com/jcabi/github/RtStatusTest.java (hunk 5 above)
Comment (cvrebert) [inline src/test/java/com/jcabi/github/RtStatusTest.java]: Done.

File: src/main/java/com/jcabi/github/Status.java (hunk 4 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/Status.java]: Also, that's presumably slower since it would redo the lowercasing every time the method is invoked (as opposed to just returning a constant instance var).

File: src/main/java/com/jcabi/github/Status.java (hunk 3 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/Status.java]: @caarlos0 Reluctantly done.

**Comment by cvrebert** [comment]
@caarlos0 Responded to all your comments.

File: src/main/java/com/jcabi/github/Status.java (hunk 4 above)
Comment (caarlos0) [inline src/main/java/com/jcabi/github/Status.java]: @cvrebert yeah, that makes sense...

**Comment by caarlos0** [comment]
@rultor merge

File: src/main/java/com/jcabi/github/RtStatus.java (hunk 1 above)
Comment (yegor256) [inline src/main/java/com/jcabi/github/RtStatus.java]: @cvrebert @caarlos0 we can't use `Optional

**Comment by yegor256** [comment]
@cvrebert looks like this pull request is against the entire design of jcabi-github... we must have all objects immutable and their methods must return values of fail with exceptions. These changes are making our classes mutable. That's a wrong idea. See http://www.yegor256.com/2014/06/09/objects-should-be-immutable.html

File: src/main/java/com/jcabi/github/RtStatus.java (hunk 1 above)
Comment (cvrebert) [inline src/main/java/com/jcabi/github/RtStatus.java]: @yegor256 Can you please elaborate? You mean because the generic T might be mutable?

**Comment by cvrebert** [comment]
@yegor256 WTF are you talking about? `Optional<T>` is immutable provided that T is immutable, which it is here since they're Strings.

**Comment by yegor256** [comment]
@cvrebert the design of Status/RtStatus is already different from all other classes/interfaces we have here. We should not return `description()` (for example). Instead, we should return only `json()` and then create `Status.Smart`, which will have `hasDescription()` and `description()`. make sense?

**Comment by cvrebert** [comment]
@yegor256 Okay, then say that in the first place instead of talking nonsense about immutability :-) I'm willing to add some commits to rework the API to fix the design.

**Comment by yegor256** [comment]
@cvrebert well, the immutability problem is also there :) check this: http://aspects.jcabi.com/annotation-immutable.html

> Keep in mind that, although JDK requires some classes to be immutable, most of them are mutable in reality. For example, BigDecimal. (Well, at least in OpenJDK 7). Our immutability checker is rather strict and doesn't allow classes to have any non-private, non-final properties in order to be called truly immutable.

So, for jcabi-aspects class `Optional` is NOT immutable

**Comment by cvrebert** [comment]
@yegor256 Submitted new PR that fixes the design of the Status class like you mentioned, along with other status API fixes: #1127

**Comment by caarlos0** [comment]
@dmarkov please take a look here...

**Comment by dmarkov** [comment]
@caarlos0 Thank you, I have added **18 mins** to your account in payment/transaction "`AP-593458880N415643L`", time consumed: 267 hours and 38 mins; review comments (c=3) added as a bonus; +18 to your rating, your total score is [+2326](http://www.netbout.com/b/35095?open=rating)
